### ISSUE TITLE
QueryOptions.SkipTake works with native Umbraco Examine Lucence, but not with ExamineX Elasticsearch

### ISSUE THREAD

**Issue by raimondkempees-arc** [SDKits/ExamineX#114]
**Describe the bug**
QueryOptions.SkipTake works with native Umbraco Examine Lucence, but not with ExamineX Elasticsearch

**To Reproduce**

Execute a simple query that gives back more results than _searchResultsPerPage (which is 12)
var results = query.Execute(
                    QueryOptions.SkipTake(
                        0,
                        _searchResultsPerPage)
                );

For the Lucene Searcher I get back 12 results and the skip is respected.

For the Elastic based searcher I get back all results.

**Expected behavior**
The expectation is that for both search engines SkipTake works the same - or there is documentation or guidance that shows how to do it differently for Elastic.

**Versions**
 - ExamineX version - 6.0.8
 - ExamineX.ElasticSearch.Umbraco - 3.0.6
 - Umbraco version - 13.5.2
 - Examine version - 3.3.0

**Comment by Shazwazza** [comment]
Thanks for the report. I'll check this out today. There's definitely tests that cover this so will get back to you once I boot up this morning.

**Comment by Shazwazza** [comment]
Hi @raimondkempees-arc 

One of the simple SkipTake tests in the ExamineX suite is this one:

```cs
public void Skip_Take()
{
    using (_examineXIndex.ProcessNonAsync())
    {
        _index.IndexItems(new[] {
            ValueSet.FromObject("1-" + NewId(), "content",
                new { Content = "mystic world" }),
            ValueSet.FromObject("2-" + NewId(), "content",
                new { Content = "mystic worlds" }),
            ValueSet.FromObject("3-" + NewId(), "content",
                new { Content = "mystic you cruel world" }),
            ValueSet.FromObject("4-" + NewId(), "content",
                new { Content = "hi there, mystic world" })
        });
    }

    var searcher = _getSearcher();

    var criteria = searcher.CreateQuery();
    var filter = criteria.Field("Content", "mystic");

    //Act
    var result = filter.Execute(QueryOptions.SkipTake(0, 3));
    var results = result.ToArray();

    //Assert

    Assert.AreEqual(3, results.Length);

    //NOTE: These are the total matched! The actual results are limited
    Assert.AreEqual(4, result.TotalItemCount);
}
```

This works the same on Examine and ExamineX for both Azure and Elastic Search.

I've adjusted it to have more items and using the value of 12 that you are:

```cs
public void Skip_Take()
{
    using (_examineXIndex.ProcessNonAsync())
    {
        var valueSets = new List<ValueSet>();
        for (var i = 0; i < 100; i++)
        {
            valueSets.Add(
                ValueSet.FromObject(i + "-" + NewId(), "content", new { Content = "mystic world" }));
        }

        _index.IndexItems(valueSets);
    }

    var searcher = _getSearcher();

    var criteria = searcher.CreateQuery();
    var filter = criteria.Field("Content", "mystic");

    //Act
    var result = filter.Execute(QueryOptions.SkipTake(0, 12));
    var results = result.ToArray();

    //Assert

    Assert.AreEqual(12, results.Length);

    //NOTE: These are the total matched! The actual results are limited
    Assert.AreEqual(100, result.TotalItemCount);
}
```

Perhaps you can share more of the code you are using to see if there is some nuance with something specific that is operating differently from the above.

Thanks!

**Comment by raimondkempees-arc** [comment]
Thanks for looking into it. I do this:

```
if (_examineManager.TryGetIndex(UmbracoConstants.UmbracoIndexes.ExternalIndexName, out var examineIndex))
        {
            if (examineIndex is not null)
            {
                var searcher = examineIndex.Searcher;
                var query = searcher.CreateQuery("content")
                    .ManagedQuery(q,
                    [
                        "content", "metaTitle", "metaDescription", "metaKeywords", "heading", "articleContent",
                        "quote", "blocks", "fileTextContent"
                    ]);

                var results = query.Execute(
                    // Skip 12*3 = skip the first 36 results
                    // Take 12 = the results up to 48, creating 48-36 results
                    
                    // TODO: there's a bug in ExamineX Elastic that doesn't respect SkipTake
                    QueryOptions.SkipTake(
                        0, 12)
                );

                var totalResultCount = results.TotalItemCount;
                var resultsInResultSet = results.ToArray().Length;

                _logger.LogDebug(
                    "We have a total {numberOfResults} results, with in the resultSet: {resultsInResultSet} items .",
                    totalResultCount, resultsInResultSet);
}
```

So, I'll try and get it working with using ManagedQuery - although locally on Lucene this works completely fine.

**Comment by Shazwazza** [comment]
Great thx for that. If I can replicate today I'll get a fix out asap and add to the test suite.

**Comment by raimondkempees-arc** [comment]
Great  news! Thanks.

On Dec 13, 2024, at 17:56, Shannon Deminick ***@***.***> wrote:

﻿

Hi @raimondkempees-arc<https://github.com/raimondkempees-arc>,

I can replicate with the latest published package but I cannot replicate from source or build pipelines which is very unusual. The good news at least is that I can replicate so should be able to get a fix out asap. I'll let you know once its out.

**Comment by raimondkempees-arc** [comment]
By the way – the Skip part does work, it’s the Take that doesn’t.

From: Raimond Kempees ***@***.***>
Date: Friday, 13 December 2024 at 18:02
To: SDKits ExamineX ***@***.***>
Cc: SDKits ExamineX ***@***.***>, Mention ***@***.***>
Subject: Re: [SDKits/ExamineX] QueryOptions.SkipTake works with native Umbraco Examine Lucence, but not with ExamineX Elasticsearch (Issue #114)
Great  news! Thanks.


On Dec 13, 2024, at 17:56, Shannon Deminick ***@***.***> wrote:
﻿

Hi @raimondkempees-arc<https://github.com/raimondkempees-arc>,

I can replicate with the latest published package but I cannot replicate from source or build pipelines which is very unusual. The good news at least is that I can replicate so should be able to get a fix out asap. I'll let you know once its out.

**Comment by Shazwazza** [comment]
It is an issue with obfuscation and NEST APIs. I have it fixed but the challenge is ensuring the build is now running all test suites against both debug and release versions in both windows and Linux. Should have a new version out on Monday. Thanks for your info and patience!

**Comment by raimondkempees-arc** [comment]
Thank you for fixing! ☺️

On Dec 14, 2024, at 16:37, Shannon Deminick ***@***.***> wrote:

﻿

It is an issue with obfuscation and NEST APIs. I have it fixed but the challenge is ensuring the build is now running all test suites against both debug and release versions in both windows and Linux. Should have a new version out on Monday. Thanks for your info and patience!

**Comment by Shazwazza** [comment]
Hi @raimondkempees-arc,

This is now fixed and version 3.0.9 has been pushed to Nuget. Tests suites are now all run against both Debug and Release and Obfuscated builds to avoid potential issues like this in the future.

**Comment by raimondkempees-arc** [comment]
Hi,

I’m afraid I get the exact same issue. Updated NuGet versions:

***@***.***


Debug screenshot:

***@***.***

Am I doing something wrong?

Thanks,

Raimond

From: Shannon Deminick ***@***.***>
Date: Monday, 16 December 2024 at 23:24
To: SDKits/ExamineX ***@***.***>
Cc: Raimond Kempees ***@***.***>, Mention ***@***.***>
Subject: Re: [SDKits/ExamineX] QueryOptions.SkipTake works with native Umbraco Examine Lucence, but not with ExamineX Elasticsearch (Issue #114)

Hi @raimondkempees-arc<https://github.com/raimondkempees-arc>,

This is now fixed and version 3.0.9 has been pushed to Nuget. Tests suites are now all run against both Debug and Release and Obfuscated builds to avoid potential issues like this in the future.

**Comment by raimondkempees-arc** [comment]
Screenshots:

<img width="543" alt="Screenshot 2024-12-18 at 11 23 38" src="https://github.com/user-attachments/assets/dfff7348-f308-4ad2-9cfe-e61efb7146ea" />

<img width="1009" alt="Screenshot 2024-12-18 at 11 22 22" src="https://github.com/user-attachments/assets/df7f66da-b7d9-48d6-975c-d4ed35489650" />

**Comment by Shazwazza** [comment]
Hi @raimondkempees-arc,

Turns out this issue occurs when unlicensed. I've create a bug here that you can track: https://github.com/SDKits/ExamineX/issues/115

Fix will be out later today.

**Comment by raimondkempees-arc** [comment]
Ok, thanks!

On Dec 18, 2024, at 17:44, Shannon Deminick ***@***.***> wrote:

﻿

Hi @raimondkempees-arc<https://github.com/raimondkempees-arc>,

Turns out this issue occurs when unlicensed. I've create a bug here that you can track: #115<https://github.com/SDKits/ExamineX/issues/115>

Fix will be out later today.

**Comment by Shazwazza** [comment]
@raimondkempees-arc the fix for #115 has been shipped with a new 3.0.10 version

**Comment by raimondkempees-arc** [comment]
Awesome thanks!

From: Shannon Deminick ***@***.***>
Date: Friday, 20 December 2024 at 17:15
To: SDKits/ExamineX ***@***.***>
Cc: Raimond Kempees ***@***.***>, Mention ***@***.***>
Subject: Re: [SDKits/ExamineX] QueryOptions.SkipTake works with native Umbraco Examine Lucence, but not with ExamineX Elasticsearch (Issue #114)

@raimondkempees-arc<https://github.com/raimondkempees-arc> the fix for #115<https://github.com/SDKits/ExamineX/issues/115> has been shipped with a new 3.0.10 version

**Comment by raimondkempees-arc** [comment]
This is confirmed fixed.

From: Raimond Kempees ***@***.***>
Date: Monday, 23 December 2024 at 12:20
To: SDKits/ExamineX ***@***.***>, SDKits/ExamineX ***@***.***>
Cc: Mention ***@***.***>
Subject: Re: [SDKits/ExamineX] QueryOptions.SkipTake works with native Umbraco Examine Lucence, but not with ExamineX Elasticsearch (Issue #114)
Awesome thanks!

From: Shannon Deminick ***@***.***>
Date: Friday, 20 December 2024 at 17:15
To: SDKits/ExamineX ***@***.***>
Cc: Raimond Kempees ***@***.***>, Mention ***@***.***>
Subject: Re: [SDKits/ExamineX] QueryOptions.SkipTake works with native Umbraco Examine Lucence, but not with ExamineX Elasticsearch (Issue #114)

@raimondkempees-arc<https://github.com/raimondkempees-arc> the fix for #115<https://github.com/SDKits/ExamineX/issues/115> has been shipped with a new 3.0.10 version
